### DISCUSSION TITLE
使用linq.dynamic生成的expression后报错

### DISCUSSION THREAD

**Discussion by tiansfather** [dotnetcore/FreeSql#1378] [General]
错误信息是：variable '' of type 'MyEntity' referenced from scope '', but it is not defined
有什么办法可以支持？

`using FreeSql;
using FreeSql.DataAnnotations;
using System;
using System.Linq.Dynamic.Core;
using System.Linq.Expressions;

var fsql = new FreeSqlBuilder()
                 .UseConnectionString(DataType.MySql, "Server=localhost;Database=db;charset=utf8;user=root;password=root")
                 .UseAutoSyncStructure(true)
                 .Build();

Expression<Func<MyEntity, string>> newexp = x => x.Name;

var func = DynamicExpressionParser.ParseLambda<MyEntity, bool>(new ParsingConfig(), true, "@0(it)==\"aa\"", new[] { newexp });

//报错variable '' of type 'MyEntity' referenced from scope '', but it is not defined
var a = fsql.Select<MyEntity>().Where(func).Count();

internal class MyEntity
{
    [Column(IsIdentity = true)]
    public int Id { get; set; }

    public string Name { get; set; }
}`

**Comment by 2881099** [comment]
func 具体类型是什么，https://github.com/dotnetcore/FreeSql/issues/1113 这个可能有帮助

**Comment by tiansfather** [comment]
func 就是Expression<Func<MyEntity,bool>>类型，具体错误堆栈如下：
在 System.Linq.Expressions.Compiler.VariableBinder.Reference(ParameterExpression node, VariableStorageKind storage)
   在 System.Linq.Expressions.Compiler.VariableBinder.VisitParameter(ParameterExpression node)
   在 System.Linq.Expressions.ParameterExpression.Accept(ExpressionVisitor visitor)
   在 System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   在 System.Linq.Expressions.Compiler.VariableBinder.VisitInvocation(InvocationExpression node)
   在 System.Linq.Expressions.InvocationExpression.Accept(ExpressionVisitor visitor)
   在 System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   在 System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   在 System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
   在 System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   在 System.Linq.Expressions.Compiler.VariableBinder.Visit(Expression node)
   在 System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda)
   在 System.Linq.Expressions.LambdaExpression.Compile()
   在 FreeSql.Internal.CommonExpression.ExpressionLambdaToSql(Expression exp, ExpTSC tsc) 在 C:\Users\28810\Desktop\github\FreeSql\FreeSql\Internal\CommonExpression.cs 中: 第 926 行
![image](https://user-images.githubusercontent.com/16606743/208559621-878c3980-1d79-472a-922e-518c1276e118.png)
![image](https://user-images.githubusercontent.com/16606743/208560657-1856dbd8-dc3c-4fc7-aa2c-5ae6e1806bea.png)

这种方式生成的expression在ef core中是可以使用的，但在freesql中好像解析不了。

测试源码在https://gitee.com/tiansfather/fortest.git，您有空了麻烦帮忙看看

 DynamicFilterInfo 还没用过，我去看看

**Comment by 2881099** [comment]
fsql.Aop.ParseExpression 解决办法如下：（v3.2.685 版本会集成）

```c#
using FreeSql;
using FreeSql.DataAnnotations;
using System;
using System.Linq.Dynamic.Core;
using System.Linq.Expressions;

var fsql = new FreeSqlBuilder()
                 .UseConnectionString(DataType.Sqlite, "data source=:memory:")
                 .UseAutoSyncStructure(true)
                 .UseNoneCommandParameter(true)
                 .UseMonitorCommand(cmd => Console.WriteLine(cmd.CommandText))
                 .Build();

fsql.Aop.ParseExpression += (_, e) =>
{
    if (e.Expression.NodeType == ExpressionType.Invoke && 
        e.Expression is InvocationExpression invokeExp)
    {
        var replaceVistor = new FreeSql.Internal.CommonExpression.ReplaceVisitor();
        var replaceExp = invokeExp.Expression;
        var lambdaExp = replaceExp as LambdaExpression;
        var len = Math.Min(invokeExp.Arguments.Count, lambdaExp.Parameters.Count);
        for (var a = 0; a < len; a++)
            replaceExp = replaceVistor.Modify(replaceExp, lambdaExp.Parameters[a], invokeExp.Arguments[a]);
        e.Result = e.FreeParse(replaceExp);
    }
};

//定义一个expression
Expression<Func<MyEntity, string>> newexp = x => x.Name;

//使用System.Linq.Dynamic.Core 包装成另一个expression
var func = DynamicExpressionParser.ParseLambda<MyEntity, bool>(new ParsingConfig(), true, "@0(it)==\"aa\"", new[] { newexp });

//报错variable '' of type 'MyEntity' referenced from scope '', but it is not defined
var a = fsql.Select<MyEntity>().Where(func).Count();

internal class MyEntity
{
    [Column(IsIdentity = true)]
    public int Id { get; set; }

    public string Name { get; set; }
}
```

**Reply by 2881099** [reply]
v3.2.685 已发布

**Reply by tiansfather** [reply]
感谢
